R/match_on_ltcs.R

```diff
@@ -0,0 +1,100 @@
+      chi_dob_max = phsmethods::dob_from_chi(
+        chi_number = chi,
+        chi_check = FALSE,
+        min_date = lubridate::ymd("2000-01-01"),
+        max_date = end_fy(year)
+      ),
+      # Compute two ages for each chi, the maximum and minimum it could be
+      chi_age_max = compute_mid_year_age(year, chi_dob_min),
+      chi_age_min = compute_mid_year_age(year, chi_dob_max),
+      # Mutate age based on different scenarios
+      age = dplyr::case_when(
+        # Case when the dob is already valid
+        .data$dob == chi_dob_min ~ chi_age_max,
+        .data$dob == chi_dob_max ~ chi_age_min,
+        # If one option for dob isn't there, use the other
+        is.na(chi_dob_min) &
+          !is.na(chi_dob_max) ~ chi_age_min, !is.na(chi_dob_min) &
+          is.na(chi_dob_max) ~ chi_age_max,
+        # If they have an LTC date before birth date, assume older
+        chi_dob_max > purrr::reduce(dplyr::select(., "arth_date":"digestive_date"), `min`) ~ chi_age_max,
+        # If they have a GLS record and the age is broadly correct, assume older
+        dplyr::between(chi_age_max, 50, 130) &
+          recid == "GLS" ~ chi_age_max,
+        # If a congenital defect lines up with a dob, assume it is correct
+        chi_dob_max == congen_date ~ chi_age_min,
+        chi_dob_min == congen_date ~ chi_age_max,
+        # If being older makes them over 113, assume they are younger
+        chi_age_max > 113 ~ chi_age_min
+      ),
+      # For the same scenarios, mutate dob
+      dob = dplyr::case_when(
+        # Case when the dob is already valid
+        .data$dob == chi_dob_min ~ dob,
+        .data$dob == chi_dob_max ~ dob,
+        # If one option for dob isn't there, use the other
+        is.na(chi_dob_min) &
+          !is.na(chi_dob_max) ~ chi_dob_max, !is.na(chi_dob_min) &
+          is.na(chi_dob_max) ~ chi_dob_min,
+        # If they have an LTC date before birth date, assume older
+        chi_dob_max > purrr::reduce(dplyr::select(., "arth_date":"digestive_date"), `min`) ~ chi_dob_min,
+        # If they have a GLS record and the age is broadly correct, assume older
+        dplyr::between(chi_age_max, 50, 130) &
+          recid == "GLS" ~ chi_dob_min,
+        # If a congenital defect lines up with a dob, assume it is correct
+        chi_dob_max == congen_date ~ chi_dob_max,
+        chi_dob_min == congen_date ~ chi_dob_min,
+        # If being older makes them over 113, assume they are younger
+        chi_age_max > 113 ~ chi_dob_max
+      )
+    ) %>%
+    # If we still don't have an age, try and fill it in from other records.
+    dplyr::group_by(chi) %>%
+    tidyr::fill(dob, .direction = "downup") %>%
+    dplyr::ungroup() %>%
+    # Fill in ages for any that are left.
+    dplyr::mutate(age = dplyr::if_else(is.na(age), compute_mid_year_age(year, dob), age)) %>%
+    # If any gender codes are missing or 0 recode to CHI gender.
+    dplyr::mutate(chi_gender = phsmethods::sex_from_chi(chi))
```

Comment:
We need to do something else with this e.g. 
```suggestion
    dplyr::mutate(chi_gender = phsmethods::sex_from_chi(chi),
    gender = dplyr::if_else(!is.na(chi_gender) & (is_missing(gender) | gender == 0L), chi_gender, gender)
```

Reply:
Yeah. Looks good. I also deleted the columns "chi_age_max", etc. 

Just another reflection here. `if_else` requires type stable, meaning values for condtion true or false should be of the same type. In this case, gender was imported as numeric which chi_gender is integer. So it reported an error in `if_else`. Subsequently, I do `mutate(gender = as.integer(gender))` at first. By the way, `base::ifelse` does not require type stable.